Add Carousel with an example

Calvicii opened the pull request:
Added the Carousel widget


https://github.com/user-attachments/assets/3a9f9912-5134-4086-9b80-7bb9bc877b2f

MMarco94 commented on src/main/kotlin/io/github/mmarco94/compose/adw/components/Carousel.kt:
```diff
@@ -0,0 +1,66 @@
+            set(revealDuration) { this.widget.revealDuration = it }
+            set(scrollParams) { this.widget.scrollParams = it }
+            set(spacing) { this.widget.spacing = it }
+            set(onPageChanged) { this.widget.onPageChanged { onPageChanged(it) } }
```
The previous `SignalConnection` needs to be disconnected. 
You can achieve this by saving the latest SignalConnection on the `GtkContainerComposeNode`.

You can look at `AdwBannerComposeNode`'s `onButtonClicked` as an example (https://github.com/MMarco94/compose-4-gtk/blob/81baa5af33c284cbe628f731fd055ecfbe99fb1a/src/main/kotlin/io/github/mmarco94/compose/adw/components/Banner.kt#L22)

MMarco94 commented on src/test/kotlin/Carousel.kt:
```diff
@@ -0,0 +1,29 @@
+fun main(args: Array<String>) {
+    application("my.example.HelloApp", args) {
+        ApplicationWindow("Test", onClose = ::exitApplication) {
+            Box(orientation = Orientation.VERTICAL) {
```
Optional: Can you add some ToggleButtons to test the various parameters of Carousel (e.g. `allowLongSwipes`)?

MMarco94 commented on src/main/kotlin/io/github/mmarco94/compose/adw/components/Carousel.kt:
```diff
@@ -0,0 +1,66 @@
+    scrollParams: SpringParams = SpringParams(1.0, 0.5, 500.0),
+    spacing: Int = 0,
+    onPageChanged: (Int) -> Unit = {},
+    content: @Composable () -> Unit,
```
Question: does it make sense to change the interface of this function, and pass it the number of pages; change `content` to get an int as input (the page index); and do the for loop inside this function?

I'm asking because that's the interface of an HorizontalPager on the "normal" compose (https://developer.android.com/develop/ui/compose/layouts/pager). We don't have to mimic the same interface, but it's worth raising the question.

MMarco94 commented:
I've left some comments (only one mandatory).

Thanks!

Calvicii commented:
After "studying" the android doc you sent, I decided to follow their way of doing things. I've added a CarouselState class that looks like this:

``` Kotlin
class CarouselState(val pageCount: Int) {
    private var _currentPage by mutableStateOf(0)

    var currentPage: Int
        get() = _currentPage
        set(value) {
            if (value in 0 until pageCount) {
                _currentPage = value
            }
        }
}
```

The carousel state is required in order to make a Carousel.

``` Kotlin
val pageState = remember { CarouselState(8) }

Carousel(
    state = pageState,
    ...
) {
...
}
```

I prefer this over my old code because this uses the Carousel.scrollTo() method if you change CarouselState.currentPage and scrolls to the desired page.
``` Kotlin
Button("Scroll to page 4") {
    pageState.currentPage = 4
}
```

I've also added `animateScrollTo` 
![image](https://github.com/user-attachments/assets/58427135-62da-44fb-9c16-9a6999e72c01)
which enables/disables the animation of scrolling to a page programmatically. I haven't found any mention of it's default value in the libadwaita doc so I decided to make it `true`. I've also added a bunch of tests and buttons in the example.

https://github.com/user-attachments/assets/ca5ebc35-b17e-4169-96f4-da1169023db6

MMarco94 commented:
Thanks for the changes!

I've added some commits on top: can you take a look? If they look good to you, I'll merge.

The main difference is that I added the `CarouselState.scrollTo` function, which I think streamlines how to change pages.  The previous approach used the `currentPage` State as a signal to perform an action, which is an anti-pattern.

Calvicii commented:
I might be lost but I can't find your commits. Either way, `CarouselState.scrollTo` sounds cool to me so feel free to merge!
